compiler/rustc_mir_transform/src/jump_threading.rs

```diff
@@ -492,19 +484,20 @@ impl<'a, 'tcx> TOFinder<'a, 'tcx> {
                     }
                 }
             }
-            // Transfer the conditions on the copy rhs, after inversing polarity.
+            // Transfer the conditions on the copy rhs, after inverting the value of the condition.
             Rvalue::UnaryOp(UnOp::Not, Operand::Move(place) | Operand::Copy(place)) => {
-                if !place.ty(self.body, self.tcx).ty.is_bool() {
-                    // Constructing the conditions by inverting the polarity
-                    // of equality is only correct for bools. That is to say,
-                    // `!a == b` is not `a != b` for integers greater than 1 bit.
-                    return;
-                }
+                let layout = self.ecx.layout_of(place.ty(self.body, self.tcx).ty).unwrap();
                 let Some(conditions) = state.try_get_idx(lhs, &self.map) else { return };
                 let Some(place) = self.map.find(place.as_ref()) else { return };
-                // FIXME: I think This could be generalized to not bool if we
-                // actually perform a logical not on the condition's value.
-                let conds = conditions.map(self.arena, Condition::inv);
+                let conds = conditions.map(self.arena, |mut cond| {
+                    cond.value = self
+                        .ecx
+                        .unary_op(UnOp::Not, &ImmTy::from_scalar_int(cond.value, layout))
+                        .unwrap()
+                        .to_scalar_int()
+                        .unwrap();
```

Comment:
`unwrap` is dangerous in optimizations, you probably want ti ignore errors rather than turn them into panics. So it'd be something like
```rust
if let Some(result) = self
                        .ecx
                        .unary_op(UnOp::Not, &ImmTy::from_scalar_int(cond.value, layout)).discard_err()
    && let Some(result) = result.to_scalar_int().discard_err() {
```
or maybe an `and_then` chain, seeing as this is currently in a `map`.

Reply:
If I understand the flow of the code right, failing to modify the condition value would result in miscompiles similar to the original issue. I _think_ the only way to handle a failure here would be to bail out of optimizing this body? I could be wrong there

Reply:
Yes, generally that's what happens in optimizations.

They are often written as functions returning `Option` so that one can easily do `discard_err()?`.